### PR TITLE
raftstore: support consistency check

### PR DESCRIPTION
**Pull request by BusyJay**
ref #1123 #1286 #1314

@siddontang @hhkbp2 PTAL

### CODE REVIEW

**Comment by BusyJay** [comment]
PTAL

**Comment by BusyJay** [comment]
ping

File: src/raftstore/store/metrics.rs [hunk 1]
```diff
@@ -92,4 +92,11 @@ lazy_static! {
                        2097152.0, 4194304.0, 8388608.0, 16777216.0] ]
             }
         ).unwrap();
+
+    pub static ref REGION_HASH_COUNTER_VEC: CounterVec =
+        register_counter_vec!(
+            "tikv_raftstore_hash_total",
+            "Total number of raftstore has computing.",
```
Comment (siddontang) [inline src/raftstore/store/metrics.rs]: has been computing

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -165,6 +171,13 @@ fn notify_region_removed(region_id: u64, peer_id: u64, mut cmd: PendingCmd) {
     cmd.call(resp);
 }
 
+pub struct ConsistencyState {
+    pub last_check_time: Instant,
+    // (computed_result_or_to_be_revified, index, hash)
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: what's revified, typo?

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -1568,6 +1580,189 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+        if expected_index < state.index {
+            REGION_HASH_COUNTER_VEC.with_label_values(&["verify", "miss"]).inc();
+            warn!("[region {}] has scheduled a new hash: {} > {}, skip.",
+                  region_id,
+                  state.index,
+                  expected_index);
+            return;
+        }
+
+        if state.index == expected_index {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: I think we can extract a function for both here and L1706.

File: src/raftstore/store/worker/consistency_check.rs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+                return;
+            }
+            Ok(Some(v)) => digest.write(&v),
+            Ok(None) => digest.write(b""),
+        }
+        let sum = digest.sum32();
+        timer.observe_duration();
+
+        let mut checksum = Vec::with_capacity(4);
+        checksum.write_u32::<BigEndian>(sum).unwrap();
```
Comment (BusyJay) [inline src/raftstore/store/worker/consistency_check.rs]: Yes, to reduce the overhead.

**Comment by siddontang** [comment]
Rest LGTM

**Comment by siddontang** [comment]
@disksing 

We should add MVCC verify later.

File: src/raftstore/store/config.rs [hunk 5]
```diff
@@ -99,6 +102,9 @@ pub struct Config {
     pub max_leader_missing_duration: Duration,
 
     pub snap_apply_batch_size: usize,
+
+    // Interval (s) to check region whether the data are consistent.
```
Comment (hhkbp2) [inline src/raftstore/store/config.rs]: the data s/are/is/

File: src/raftstore/store/metrics.rs [hunk 6]
```diff
@@ -92,4 +92,11 @@ lazy_static! {
                        2097152.0, 4194304.0, 8388608.0, 16777216.0] ]
             }
         ).unwrap();
+
+    pub static ref REGION_HASH_COUNTER_VEC: CounterVec =
+        register_counter_vec!(
+            "tikv_raftstore_hash_total",
+            "Total number of raftstore has computed.",
```
Comment (hhkbp2) [inline src/raftstore/store/metrics.rs]: Total number of hash ...?

File: src/raftstore/store/peer.rs [hunk 7]
```diff
@@ -187,13 +197,17 @@ pub struct Peer {
...
 
     pub tag: String,
 
     pub last_compacted_idx: u64,
 
+    // if we remove ourself in ChangePeer remove, we should set this flag, then
+    // any following committed logs in same Ready should be applied failed.
+    pending_remove: bool,
     leader_lease_expired_time: Option<Timespec>,
+    leader_missing_time: Option<Instant>,
```
Comment (hhkbp2) [inline src/raftstore/store/peer.rs]: related fields should be grouped together, such as `leader_lease_expired_time` and `election_timeout`.

File: src/raftstore/store/peer.rs [hunk 8]
```diff
@@ -1715,3 +1743,38 @@ fn make_transfer_leader_response() -> RaftCmdResponse {
...
+    fn exec_compute_hash(&self,
+                         ctx: &ExecContext,
+                         _: &AdminRequest)
+                         -> Result<(AdminResponse, Option<ExecResult>)> {
+        let resp = AdminResponse::new();
+        Ok((resp,
+            Some(ExecResult::ComputeHash {
+            region: self.region().clone(),
+            index: ctx.index,
+            // This snapshot may be hold for a long time, which may cause too many
```
Comment (hhkbp2) [inline src/raftstore/store/peer.rs]: s/hold/held/

**Comment by hhkbp2** [comment]
Rest LGTM

File: src/raftstore/store/store.rs [hunk 9]
```diff
@@ -1446,6 +1458,178 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+    if state.index == expected_index {
+        if state.hash != expected_hash {
+            panic!("[region {}] hash at {} not correct, want {}, got {}!!!",
+                   region_id,
+                   state.index,
+                   escape(&expected_hash),
+                   escape(&state.hash));
+        }
+        REGION_HASH_COUNTER_VEC.with_label_values(&["verify", "matched"]).inc();
+        state.hash = vec![];
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: `state.index` does not need to be updated here.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**